### DISCUSSION TITLE
Run the Extension Inside the Worker

### DISCUSSION THREAD

**Discussion by cemalgnlts** [CodinGame/monaco-vscode-api#303] [Q&A] [answered]
Hi,

I want to run an extension in `Worker`, what is the right way to do it.

Does using `ExtensionHostKind.LocalWebWorker` or `ExtensionHostKind.LocalProcess` make a difference?

**Comment by CGNonofr** [comment, marked as answer]
Yes, the first one will run in the worker, the other one in the browser directly. The registerExtension parameter allows to force it but it's not mandatory.

Note that for the worker you'll need to provide the worker configuration to the extension service override, and the latter requires to import 'localExtensionHost'

**Reply by cemalgnlts** [reply]
How should I give the file to run in `Worker`.

**index.ts**
```js
import { ExtensionHostKind, registerExtension } from "vscode/extensions";

const manifest = {
  name: "esbuild",
  publisher: "JSIDE",
  version: "1.0.0",
  engines: {
    vscode: "*",
  },
};

registerExtension(manifest, ExtensionHostKind.LocalWebWorker);

async function activation() {
  console.log(self);
  // new CrossOriginWorker("worker.ts", { type: "module" })
}

export default activation;
```

**worker.ts**
```js
console.log(self);
```

**Reply by CGNonofr** [reply]
You need to put the extension code in a separate file, register it using the 'registerFileUri' returned by 'registerExtension' and use the manifeste 'browser' property to reference it

**Reply by cemalgnlts** [reply]
Like that?
```js
import { ExtensionHostKind, IExtensionManifest, registerExtension } from "vscode/extensions";

const manifest: IExtensionManifest = {
  name: "esbuild",
  publisher: "JSIDE",
  version: "1.0.0",
  browser: "/worker.js",
  engines: {
    vscode: "*",
  },
};

const { registerFileUrl, whenReady } = registerExtension(manifest, ExtensionHostKind.LocalWebWorker);

registerFileUrl("/worker.js", new URL("./worker.js", import.meta.url).toString());
```

**Reply by CGNonofr** [reply]
Correct!

**Reply by cemalgnlts** [reply]
Unfortunately, it didn't work. I must have done something wrong.

First I include all extensions in this line *(After `vscode/localExtensionHost`)*: [init.ts#L45](/cemalgnlts/jside/blob/main/src/workspace/init.ts#L45)
Here I import the extension I want to run: [extensions.ts#24](/cemalgnlts/jside/blob/main/src/workspace/extensions.ts#L24)
This is the extension I want to make it work: [index.ts](/cemalgnlts/jside/blob/main/src/extensions/esbuild/index.ts)

The codes work but the `worker.js` file is not called.

**Reply by CGNonofr** [reply]
You don't need localExtensionHost

Did you configure the extHost worker?

**Reply by cemalgnlts** [reply]
If you are talking about Worker config, yes.

https://github.com/cemalgnlts/jside/blob/main/src/workspace/init.ts#L87

```js
const fakeWorker = new FakeWorker(new URL(ExtensionHostWorkerUrl, import.meta.url), { type: "module" });

const workerConfig: WorkerConfig = {
  url: fakeWorker.url.toString(),
  options: fakeWorker.options,
};

await initializeServices({
  ...getExtensionsServiceOverride(workerConfig),
// ...
```

**Reply by CGNonofr** [reply]
Ok, for the extension to run, it needs an activation event in the manifest

**Reply by cemalgnlts** [reply]
Great now my worker code works but I can't use `import`.
```
[Extension Host] Loading code for extension JSIDE.esbuild failed: Cannot use import statement outside a module
```

```js
import { window } from "vscode";

window.showInformationMessage("Hello from extension.");
```

**Reply by CGNonofr** [reply]
Yes, VSCode extensions are commonjs. If you want ESM you'll need to transpile it

**Reply by cemalgnlts** [reply]
Now everything works.
**Thank you.**
